### ISSUE TITLE
RFI: Mutual Auth TLS

### ISSUE THREAD

**Issue by huggsboson** [linkerd/linkerd#1102]
Box has an ... interesting ... mutual auth TLS setup.  We generate the standard cert/key pair for each service. But then also generate a series of cert authorities pems, 1 for each service this service needs to connect to and 1 big one for all incoming connections.  In essence this provides a coarse grain form of service level authorization.  If a service shouldn't talk to this service it's CA won't be in this service's CA file.

I notice from the docs that linkerd largely relies on common-name for this type of behavior.  I understand the benefits of doing that, but we have backwards compatibility to maintain and a transition period where we'd want to support both if we decided to move over.  My question is: How doable is this in linkerd?  How much effort would it be to add?

**Comment by wmorgan** [comment]
@huggsboson we discussed this a bit internally today. Will update the ticket in a bit.

**Comment by olix0r** [comment]
@huggsboson i think it's reasonably straightforward to fix this, and I think we should definitely pursue more flexible TLS configurations.  We've been waiting on [finagle 6.42](https://github.com/linkerd/linkerd/pull/1126), which adds support for clientside tls configuration.  Once this merges (today or tomorrow), I think it's simply a matter of exposing the proper configuration options from linkerd. The coding work is probably fairly minor for this, though the devil is always in the details.

**Comment by chriscoomber** [comment]
Do you know when this might make it in? It looks like finagle 6.44 is soon to be released (given that the last one was about a month ago).

I'm looking at using linkerd for HTTPS uplift (and other cool things, undoubtedly) and we like mutual TLS as a simple way to do service-service authentication (by checking that the client is trusted before serving it).

**Comment by olix0r** [comment]
We're waiting on 6.44 because 6.42 had a show-stopping bug and the fix didn't make it in 6.43.  Once we're on 6.44--if there are no _new_ show-stoppers--we'll be able to dig more deeply into this feature.  So, we can hopefully start work on this soon.  We'll be able to give a better idea of when we'll be able to _ship_ this feature once we've gotten a bit further into the investigation.

We definitely want to do it soon.  There's also a chance this feature may land in linkerd-tcp before it lands in linkerd.

**Comment by chriscoomber** [comment]
That last bit is interesting, I just read through your [blog post](https://blog.buoyant.io/2017/03/29/introducing-linkerd-tcp/) and maybe linkerd-tcp is sufficient for my use case. Thanks for the info.

**Comment by elecnix** [comment]
[Finagle 6.44](https://github.com/twitter/finagle/releases/tag/finagle-6.44.0) is out.

**Comment by adleong** [comment]
Closing this in favor of #1276
